lib/widgets/editor/editor_toolbar.dart

```diff
@@ -0,0 +1,462 @@
+        ));
+  }
+
+  String get firstSelectedLine {
+    if (text.isEmpty) return '';
+    return text.substring(text.getBeginningOfTheLine(selection.start - 1),
+        text.getEndOfTheLine(selection.end));
+  }
+
+  void insertAtBeginningOfFirstSelectedLine(String s) {
+    final lines = TextLinesIterator.fromController(this)..moveNext();
+    lines.current = s + lines.current;
+    value = value.copyWith(
+      text: lines.text,
+      selection: selection.copyWith(
+        baseOffset: selection.baseOffset + s.length,
+        extentOffset: selection.extentOffset + s.length,
+      ),
+    );
+  }
+
+  void removeAtBeginningOfEverySelectedLine(String s) {
+    final lines = TextLinesIterator.fromController(this);
+    var linesCount = 0;
+    while (lines.moveNext()) {
+      if (lines.isWithinSelection) {
+        if (lines.current.startsWith(RegExp.escape(s))) {
+          lines.current = lines.current.substring(s.length);
+          linesCount++;
+        }
+      }
+    }
+
+    value = value.copyWith(
+      text: lines.text,
+      selection: selection.copyWith(
+        baseOffset: selection.baseOffset - s.length,
+        extentOffset: selection.extentOffset - s.length * linesCount,
+      ),
+    );
+  }
+
+  void insertAtBeginningOfEverySelectedLine(String s) {
+    final lines = TextLinesIterator.fromController(this);
+    var linesCount = 0;
+    while (lines.moveNext()) {
+      if (lines.isWithinSelection) {
+        if (!lines.current.startsWith(RegExp.escape(s))) {
+          lines.current = '$s${lines.current}';
+          linesCount++;
+        }
+      }
+    }
+
+    value = value.copyWith(
+      text: lines.text,
+      selection: selection.copyWith(
+        baseOffset: selection.baseOffset + s.length,
+        extentOffset: selection.extentOffset + s.length * linesCount,
+      ),
+    );
+  }
+
+  void reformat(Reformat Function(String selection) reformatter) {
+    final beg = beforeSelectionText;
+    final mid = selectionText;
+    final end = afterSelectionText;
+
+    final r = reformatter(mid);
+    value = value.copyWith(
+      text: '$beg${r.text}$end',
+      selection: selection.copyWith(
+        baseOffset: selection.baseOffset + r.selectionBeginningShift,
+        extentOffset: selection.extentOffset + r.selectionEndingShift,
+      ),
+    );
+  }
+
+  void reformatSimple(String text) =>
+      reformat((selection) => Reformat(text: text));
+}
+
+enum HeaderLevel {
+  h1(1),
+  h2(2),
+  h3(3),
+  h4(4),
+  h5(5),
+  h6(6);
+
+  const HeaderLevel(this.value);
+  final int value;
+}
+
+class Toolbar extends HookWidget {
+  final TextEditingController controller;
+  final String instanceHost;
+  final EditorToolbarStore store;
+  static const _height = 50.0;
+
+  Toolbar({
+    required this.controller,
+    required this.instanceHost,
+  }) : store = EditorToolbarStore(instanceHost);
+
+  @override
+  Widget build(BuildContext context) {
+    return MobxProvider.value(
+      value: store,
```

Comment:
Please provide with the normal MobxProvider (not `.value`). Otherwise it is never disposed.